In [1]:
import numpy as np 
import pandas as pd 

In [2]:
df_yield = pd.read_csv('yield1.csv')
df_yield.shape

(56717, 12)

In [3]:
df_yield.head()

Domain Code Domain  Area Code         Area  Element Code Element  Item Code  \
0          QC  Crops          2  Afghanistan          5419   Yield         56   
1          QC  Crops          2  Afghanistan          5419   Yield         56   
2          QC  Crops          2  Afghanistan          5419   Yield         56   
3          QC  Crops          2  Afghanistan          5419   Yield         56   
4          QC  Crops          2  Afghanistan          5419   Yield         56   

    Item  Year Code  Year   Unit  Value  
0  Maize       1961  1961  hg/ha  14000  
1  Maize       1962  1962  hg/ha  14000  
2  Maize       1963  1963  hg/ha  14260  
3  Maize       1964  1964  hg/ha  14257  
4  Maize       1965  1965  hg/ha  14400

In [4]:
df_yield.tail(10)

Domain Code Domain  Area Code      Area  Element Code Element  \
56707          QC  Crops        181  Zimbabwe          5419   Yield   
56708          QC  Crops        181  Zimbabwe          5419   Yield   
56709          QC  Crops        181  Zimbabwe          5419   Yield   
56710          QC  Crops        181  Zimbabwe          5419   Yield   
56711          QC  Crops        181  Zimbabwe          5419   Yield   
56712          QC  Crops        181  Zimbabwe          5419   Yield   
56713          QC  Crops        181  Zimbabwe          5419   Yield   
56714          QC  Crops        181  Zimbabwe          5419   Yield   
56715          QC  Crops        181  Zimbabwe          5419   Yield   
56716          QC  Crops        181  Zimbabwe          5419   Yield   

       Item Code   Item  Year Code  Year   Unit  Value  
56707         15  Wheat       2007  2007  hg/ha  29998  
56708         15  Wheat       2008  2008  hg/ha  30097  
56709         15  Wheat       2009  2009  hg/ha  30000  
56710         15  Wheat       2010  2010  hg/ha  27681  
56711         15  Wheat       2011  2011  hg/ha  26274  
56712         15  Wheat       2012  2012  hg/ha  24420  
56713         15  Wheat       2013  2013  hg/ha  22888  
56714         15  Wheat       2014  2014  hg/ha  21357  
56715         15  Wheat       2015  2015  hg/ha  19826  
56716         15  Wheat       2016  2016  hg/ha  18294

rename **Value** to **hg/ha_yield** 
    |   removing  of unnecessary coloumns like Area Code, Domain, Item Code, etc.

In [5]:
# rename columns.
df_yield = df_yield.rename(index=str, columns={"Value": "hg/ha_yield"})
df_yield.head()

Domain Code Domain  Area Code         Area  Element Code Element  Item Code  \
0          QC  Crops          2  Afghanistan          5419   Yield         56   
1          QC  Crops          2  Afghanistan          5419   Yield         56   
2          QC  Crops          2  Afghanistan          5419   Yield         56   
3          QC  Crops          2  Afghanistan          5419   Yield         56   
4          QC  Crops          2  Afghanistan          5419   Yield         56   

    Item  Year Code  Year   Unit  hg/ha_yield  
0  Maize       1961  1961  hg/ha        14000  
1  Maize       1962  1962  hg/ha        14000  
2  Maize       1963  1963  hg/ha        14260  
3  Maize       1964  1964  hg/ha        14257  
4  Maize       1965  1965  hg/ha        14400

In [6]:
# drop unwanted columns.
df_yield = df_yield.drop(['Year Code','Element Code','Element','Year Code','Area Code','Domain Code','Domain','Unit','Item Code'], axis=1)
df_yield.head()

Area   Item  Year  hg/ha_yield
0  Afghanistan  Maize  1961        14000
1  Afghanistan  Maize  1962        14000
2  Afghanistan  Maize  1963        14260
3  Afghanistan  Maize  1964        14257
4  Afghanistan  Maize  1965        14400

In [7]:
df_yield.describe()

Year     hg/ha_yield
count  56717.000000    56717.000000
mean    1989.669570    62094.660084
std       16.133198    67835.932856
min     1961.000000        0.000000
25%     1976.000000    15680.000000
50%     1991.000000    36744.000000
75%     2004.000000    86213.000000
max     2016.000000  1000000.000000

In [8]:
df_yield.info()

<class 'pandas.core.frame.DataFrame'>
Index: 56717 entries, 0 to 56716
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Area         56717 non-null  object
 1   Item         56717 non-null  object
 2   Year         56717 non-null  int64 
 3   hg/ha_yield  56717 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 2.2+ MB


### Climate Data : Rainfall 


In [9]:
df_rain = pd.read_csv('rainfall.csv')
df_rain.head()

Area  Year average_rain_fall_mm_per_year
0  Afghanistan  1985                           327
1  Afghanistan  1986                           327
2  Afghanistan  1987                           327
3  Afghanistan  1989                           327
4  Afghanistan  1990                           327

In [10]:
df_rain.tail()

Area  Year average_rain_fall_mm_per_year
6722  Zimbabwe  2013                           657
6723  Zimbabwe  2014                           657
6724  Zimbabwe  2015                           657
6725  Zimbabwe  2016                           657
6726  Zimbabwe  2017                           657

In [11]:
df_rain = df_rain.rename(index=str, columns={" Area": 'Area'})

Checking for the datatypes

In [12]:
df_rain.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6727 entries, 0 to 6726
Data columns (total 3 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Area                           6727 non-null   object
 1   Year                           6727 non-null   int64 
 2   average_rain_fall_mm_per_year  5953 non-null   object
dtypes: int64(1), object(2)
memory usage: 210.2+ KB


convert average_rain_fall_mm_per_year from object to float

In [13]:
df_rain['average_rain_fall_mm_per_year'] = pd.to_numeric(df_rain['average_rain_fall_mm_per_year'],errors = 'coerce')
df_rain.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6727 entries, 0 to 6726
Data columns (total 3 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Area                           6727 non-null   object 
 1   Year                           6727 non-null   int64  
 2   average_rain_fall_mm_per_year  5947 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 210.2+ KB


Next, droping any empty rows from dataset and merge yield dataframe with rain dataframe by year and area columns

In [14]:
df_rain = df_rain.dropna()

In [15]:
df_rain.describe()

Year  average_rain_fall_mm_per_year
count  5947.000000                    5947.000000
mean   2001.365899                    1124.743232
std       9.526335                     786.257365
min    1985.000000                      51.000000
25%    1993.000000                     534.000000
50%    2001.000000                    1010.000000
75%    2010.000000                    1651.000000
max    2017.000000                    3240.000000

The rainfall dataframe begins at 1985 and ends at 2016. 

Merge Yield Dataframe with rain Dataframe by year and area columns 

In [16]:
yield_df = pd.merge(df_yield, df_rain, on=['Year','Area'])

In [17]:
yield_df.head()

Area   Item  Year  hg/ha_yield  average_rain_fall_mm_per_year
0  Afghanistan  Maize  1985        16652                          327.0
1  Afghanistan  Maize  1986        16875                          327.0
2  Afghanistan  Maize  1987        17020                          327.0
3  Afghanistan  Maize  1989        16963                          327.0
4  Afghanistan  Maize  1990        17582                          327.0

We can see that now the years start from the first yield dataframe the starting year was 1961, now it's 1985 because that's when the rainfall data begins. 

In [18]:
yield_df.describe()

Year    hg/ha_yield  average_rain_fall_mm_per_year
count  25385.000000   25385.000000                   25385.000000
mean    2001.278787   68312.278353                    1254.849754
std        9.143915   75213.292733                     804.449430
min     1985.000000      50.000000                      51.000000
25%     1994.000000   17432.000000                     630.000000
50%     2001.000000   38750.000000                    1150.000000
75%     2009.000000   94286.000000                    1761.000000
max     2016.000000  554855.000000                    3240.000000

### Pesticides Data:
Pesticides used for each item and country was also collected from FAO database.  

In [19]:
df_pes = pd.read_csv('pesticides.csv')
df_pes.head()

Domain     Area Element                Item  Year  \
0  Pesticides Use  Albania     Use  Pesticides (total)  1990   
1  Pesticides Use  Albania     Use  Pesticides (total)  1991   
2  Pesticides Use  Albania     Use  Pesticides (total)  1992   
3  Pesticides Use  Albania     Use  Pesticides (total)  1993   
4  Pesticides Use  Albania     Use  Pesticides (total)  1994   

                           Unit  Value  
0  tonnes of active ingredients  121.0  
1  tonnes of active ingredients  121.0  
2  tonnes of active ingredients  121.0  
3  tonnes of active ingredients  121.0  
4  tonnes of active ingredients  201.0

In [20]:
df_pes = df_pes.rename(index=str, columns={"Value": "pesticides_tonnes"})
df_pes = df_pes.drop(['Element','Domain','Unit','Item'], axis=1)
df_pes.head()

Area  Year  pesticides_tonnes
0  Albania  1990              121.0
1  Albania  1991              121.0
2  Albania  1992              121.0
3  Albania  1993              121.0
4  Albania  1994              201.0

In [21]:
df_pes.describe()

Year  pesticides_tonnes
count  4349.000000       4.349000e+03
mean   2003.138883       2.030334e+04
std       7.728044       1.177362e+05
min    1990.000000       0.000000e+00
25%    1996.000000       9.300000e+01
50%    2003.000000       1.137560e+03
75%    2010.000000       7.869000e+03
max    2016.000000       1.807000e+06

In [22]:
df_pes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4349 entries, 0 to 4348
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Area               4349 non-null   object 
 1   Year               4349 non-null   int64  
 2   pesticides_tonnes  4349 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 135.9+ KB


Merge Pesticides dataframe with yield dataframe 

In [23]:
yield_df = pd.merge(yield_df, df_pes, on=['Year','Area'])
yield_df.shape

(18949, 6)

In [24]:
yield_df.head()

Area   Item  Year  hg/ha_yield  average_rain_fall_mm_per_year  \
0  Albania  Maize  1990        36613                         1485.0   
1  Albania  Maize  1991        29068                         1485.0   
2  Albania  Maize  1992        24876                         1485.0   
3  Albania  Maize  1993        24185                         1485.0   
4  Albania  Maize  1994        25848                         1485.0   

   pesticides_tonnes  
0              121.0  
1              121.0  
2              121.0  
3              121.0  
4              201.0

### Average Temprature: 

Average Temprature for each country was colleced from World Bank Data. 

In [25]:
avg_temp=  pd.read_csv('temp1.csv')

In [26]:
avg_temp.head()

year        country  avg_temp
0  1849  Côte D'Ivoire     25.58
1  1850  Côte D'Ivoire     25.52
2  1851  Côte D'Ivoire     25.67
3  1852  Côte D'Ivoire       NaN
4  1853  Côte D'Ivoire       NaN

In [27]:
avg_temp.describe()

year      avg_temp
count  71311.000000  68764.000000
mean    1905.799007     16.183876
std       67.102099      7.592960
min     1743.000000    -14.350000
25%     1858.000000      9.750000
50%     1910.000000     16.140000
75%     1962.000000     23.762500
max     2013.000000     30.730000

So average temprature starts from 1743 and ends at 2013, with some empty rows that we have to drop.

In [28]:
avg_temp = avg_temp.rename(index=str, columns={"year": "Year", "country":'Area'})
avg_temp.head()

Year           Area  avg_temp
0  1849  Côte D'Ivoire     25.58
1  1850  Côte D'Ivoire     25.52
2  1851  Côte D'Ivoire     25.67
3  1852  Côte D'Ivoire       NaN
4  1853  Côte D'Ivoire       NaN

In [29]:
yield_df = pd.merge(yield_df,avg_temp, on=['Area','Year'])
yield_df.head()

Area   Item  Year  hg/ha_yield  average_rain_fall_mm_per_year  \
0  Albania  Maize  1990        36613                         1485.0   
1  Albania  Maize  1991        29068                         1485.0   
2  Albania  Maize  1992        24876                         1485.0   
3  Albania  Maize  1993        24185                         1485.0   
4  Albania  Maize  1994        25848                         1485.0   

   pesticides_tonnes  avg_temp  
0              121.0     16.37  
1              121.0     15.36  
2              121.0     16.06  
3              121.0     16.05  
4              201.0     16.96

In [30]:
yield_df.shape

(28242, 7)

In [31]:
yield_df.describe()

Year    hg/ha_yield  average_rain_fall_mm_per_year  \
count  28242.000000   28242.000000                    28242.00000   
mean    2001.544296   77053.332094                     1149.05598   
std        7.051905   84956.612897                      709.81215   
min     1990.000000      50.000000                       51.00000   
25%     1995.000000   19919.250000                      593.00000   
50%     2001.000000   38295.000000                     1083.00000   
75%     2008.000000  104676.750000                     1668.00000   
max     2013.000000  501412.000000                     3240.00000   

       pesticides_tonnes      avg_temp  
count       28242.000000  28242.000000  
mean        37076.909344     20.542627  
std         59958.784665      6.312051  
min             0.040000      1.300000  
25%          1702.000000     16.702500  
50%         17529.440000     21.510000  
75%         48687.880000     26.000000  
max        367778.000000     30.650000

In [32]:
yield_df.isnull().sum()

Area                             0
Item                             0
Year                             0
hg/ha_yield                      0
average_rain_fall_mm_per_year    0
pesticides_tonnes                0
avg_temp                         0
dtype: int64

# Data Exploration



**yield_df** is the final obtained dataframe; 

In [33]:
yield_df.groupby('Item').count()

Area  Year  hg/ha_yield  average_rain_fall_mm_per_year  \
Item                                                                           
Cassava               2045  2045         2045                           2045   
Maize                 4121  4121         4121                           4121   
Plantains and others   556   556          556                            556   
Potatoes              4276  4276         4276                           4276   
Rice, paddy           3388  3388         3388                           3388   
Sorghum               3039  3039         3039                           3039   
Soybeans              3223  3223         3223                           3223   
Sweet potatoes        2890  2890         2890                           2890   
Wheat                 3857  3857         3857                           3857   
Yams                   847   847          847                            847   

                      pesticides_tonnes  avg_temp  
Item                                               
Cassava                            2045      2045  
Maize                              4121      4121  
Plantains and others                556       556  
Potatoes                           4276      4276  
Rice, paddy                        3388      3388  
Sorghum                            3039      3039  
Soybeans                           3223      3223  
Sweet potatoes                     2890      2890  
Wheat                              3857      3857  
Yams                                847       847

In [34]:
yield_df.describe()

Year    hg/ha_yield  average_rain_fall_mm_per_year  \
count  28242.000000   28242.000000                    28242.00000   
mean    2001.544296   77053.332094                     1149.05598   
std        7.051905   84956.612897                      709.81215   
min     1990.000000      50.000000                       51.00000   
25%     1995.000000   19919.250000                      593.00000   
50%     2001.000000   38295.000000                     1083.00000   
75%     2008.000000  104676.750000                     1668.00000   
max     2013.000000  501412.000000                     3240.00000   

       pesticides_tonnes      avg_temp  
count       28242.000000  28242.000000  
mean        37076.909344     20.542627  
std         59958.784665      6.312051  
min             0.040000      1.300000  
25%          1702.000000     16.702500  
50%         17529.440000     21.510000  
75%         48687.880000     26.000000  
max        367778.000000     30.650000

In [35]:
yield_df['Area'].nunique()

101

In [36]:
yield_df.groupby(['Area'],sort=True)['hg/ha_yield'].sum().nlargest(10)

Area
India             327420324
Brazil            167550306
Mexico            130788528
Japan             124470912
Australia         109111062
Pakistan           73897434
Indonesia          69193506
United Kingdom     55419990
Turkey             52263950
Spain              46773540
Name: hg/ha_yield, dtype: int64

In [37]:
yield_df.groupby(['Item','Area'],sort=True)['hg/ha_yield'].sum().nlargest(10)

Item            Area          
Cassava         India             142810624
Potatoes        India              92122514
                Brazil             49602168
                United Kingdom     46705145
                Australia          45670386
Sweet potatoes  India              44439538
Potatoes        Japan              42918726
                Mexico             42053880
Sweet potatoes  Mexico             35808592
                Australia          35550294
Name: hg/ha_yield, dtype: int64

In [38]:
yield_df.head()

Area   Item  Year  hg/ha_yield  average_rain_fall_mm_per_year  \
0  Albania  Maize  1990        36613                         1485.0   
1  Albania  Maize  1991        29068                         1485.0   
2  Albania  Maize  1992        24876                         1485.0   
3  Albania  Maize  1993        24185                         1485.0   
4  Albania  Maize  1994        25848                         1485.0   

   pesticides_tonnes  avg_temp  
0              121.0     16.37  
1              121.0     15.36  
2              121.0     16.06  
3              121.0     16.05  
4              201.0     16.96


### Scaling Features: 

Taking a look at the dataset above, it contains features highly varying in magnitudes, units and range. The features with high magnitudes will weigh in a lot more in the distance calculations than features with low magnitudes.

To supress this effect, we need to bring all features to the same level of magnitudes. This can be acheived by scaling.

In [39]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
features=scaler.fit_transform(features) 

NameError: name 'features' is not defined

After dropping year column in addition to scaling all values in features, the resulting array will look something like this : 

## Training Data: 
Comparing 
Linear Regression, 
Decision Tree, 
Random Forest, 
Gradient Boost, 
XGBoost, 
Bagging Regressor, 
KNeighborsRegressor, 

In [ ]:
from sklearn.model_selection import train_test_split
train_data, test_data, train_labels, test_labels = train_test_split(features, label, test_size=0.2, random_state=42)

# Model Comparison & Selection 


<a id = "basic" ></a>
# <b><span style = 'color: #24D540'> Comparing Models </span> </b>

In [ ]:
# # Importing Libraries  
# import geopandas as gpd
# import matplotlib.pyplot as plt 
# import numpy as np
# import pandas as pd
# import plotly.express as px 
# import plotly.graph_objects as go  
# from plotly.subplots import make_subplots
# import plotly.offline as py 
# import seaborn as sns
# from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, BaggingRegressor  
# from sklearn.linear_model import LinearRegression
# from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, mean_absolute_percentage_error
# from sklearn.model_selection import cross_val_score, KFold, train_test_split 
# from sklearn.neighbors import KNeighborsRegressor
# from sklearn.preprocessing import LabelEncoder, StandardScaler
# from sklearn.tree import DecisionTreeRegressor
# from xgboost import XGBRegressor

# py.init_notebook_mode(connected = True)

# sns.set_style('darkgrid')
# plt.style.use('dark_background')

In [ ]:
# import pandas as pd

# # Load the dataset
# yield_df = pd.read_csv('yield_df.csv')

# # Define the directory to save CSV files
# save_directory = r'C:\Users\SAURABH KALE\OneDrive\Documents\Design Project Sem 5\Crop Yield 2.0\csvs'

# # Iterate over unique combinations of 'Area' and 'Item'
# for area, item in yield_df[['Area', 'Item']].drop_duplicates().values:
#     # Filter the dataframe for the current combination
#     filtered_df = yield_df[(yield_df['Area'] == area) & (yield_df['Item'] == item)]
    
#     # Define the filename
#     filename = f"{area}_{item}.csv"
    
#     # Save the filtered dataframe as a CSV file
#     filtered_df.to_csv(f"{save_directory}/{filename}", index=False)
